pkg/transport/kafka/consumer.go

```diff
@@ -0,0 +1,312 @@
+			return nil, "", fmt.Errorf("failed to configure kafka-consumer - %w", err)
+		}
+	}
+
+	return kafkaConfigMap, topic, nil
+}
+
+// Consumer abstracts hub-of-hubs-kafka-transport kafka-consumer's generic usage.
+type Consumer struct {
+	log            logr.Logger
+	kafkaConsumer  *kafkaconsumer.KafkaConsumer
+	compressorsMap map[compressor.CompressionType]compressors.Compressor
+	topic          string
+
+	msgChan            chan *kafka.Message
+	bundlesUpdatesChan chan *bundle.Bundle
+
+	partitionToOffsetToCommitMap map[int32]kafka.Offset // size limited at all times (low)
+
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	startOnce  sync.Once
+	stopOnce   sync.Once
+	lock       sync.Mutex
+}
+
+// Start function starts the consumer.
+func (c *Consumer) Start() {
+	c.startOnce.Do(func() {
+		go c.handleCommits(c.ctx)
+		go c.handleKafkaMessages(c.ctx)
+	})
+}
+
+// Stop stops the consumer.
+func (c *Consumer) Stop() {
+	c.stopOnce.Do(func() {
+		c.cancelFunc()
+		close(c.msgChan)
+		c.kafkaConsumer.Close()
+	})
+}
+
+// CommitAsync commits a transported message that was processed locally.
+func (c *Consumer) CommitAsync(metadata transport.BundleMetadata) {
+	topicPartition, ok := metadata.(kafka.TopicPartition)
+	if !ok {
+		return // shouldn't happen
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if commitCandidate, found := c.partitionToOffsetToCommitMap[topicPartition.Partition]; found &&
+		topicPartition.Offset <= commitCandidate {
+		return
+	}
+
+	c.partitionToOffsetToCommitMap[topicPartition.Partition] = topicPartition.Offset
+}
+
+func (c *Consumer) handleCommits(ctx context.Context) {
+	ticker := time.NewTicker(committerInterval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C:
+			c.commitMappedOffsets()
+		}
+	}
+}
+
+func (c *Consumer) commitMappedOffsets() {
+	c.lock.Lock()
+	if len(c.partitionToOffsetToCommitMap) == 0 {
+		c.lock.Unlock()
+		return
+	}
+
+	topicPartitions := make([]kafka.TopicPartition, 0, len(c.partitionToOffsetToCommitMap))
+
+	// prepare batch for committing
+	for partition, highestOffset := range c.partitionToOffsetToCommitMap {
+		topicPartitions = append(topicPartitions, kafka.TopicPartition{
+			Topic:     &c.topic,
+			Partition: partition,
+			Offset:    highestOffset + 1, // kafka re-processes the committed offset on restart, so +1 to avoid that.
+		})
+	}
+
+	c.lock.Unlock()
+
+	if _, err := c.kafkaConsumer.Consumer().CommitOffsets(topicPartitions); err != nil {
+		c.log.Error(err, "failed to commit offsets", "TopicPartitions", topicPartitions)
+		return
+	}
+
+	// update offsets map, delete what's been committed
+	c.lock.Lock()
+	for _, topicPartition := range topicPartitions {
+		if c.partitionToOffsetToCommitMap[topicPartition.Partition] == topicPartition.Offset {
+			// no new offsets processed on this partition, delete from map
+			delete(c.partitionToOffsetToCommitMap, topicPartition.Partition)
+		}
+	}
+	c.lock.Unlock()
+
```

Comment:
same comment as in sync service..
can we do the same with concurrent queue instead of map?
you can push to the queue TopicPartition pointer.
when we dequeue all, we can arrange the data to commit (calculate highest offset).
then there is no need to do another round of comparing pointers + deleting objects from map.
(theoretically, we could just go over the topic partition and commit all of them, calculating highest in the client side is an optimization)